types/block.go

```diff
@@ -94,6 +95,25 @@ func (b *Block) ValidateBasic() error {
 	return nil
 }
 
+// IsTimely validates that the block timestamp is 'timely' according to the proposer-based timestamp algorithm.
+// To evaluate if a block is timely, its timestamp is compared to the local time of the validator along with the configured
+// Precision and MsgDelay parameters.
+// Specifically, a proposed block timestamp is considered timely if it is satisfies the following inequalities:
+//
+// proposedBlockTime > validatorLocaltime - Precision && proposedBlockTime < validatorLocalTime + Precision + MsgDelay.
+//
+// For more information on the meaning of 'timely', see the proposer-based timestamp specification:
+// https://github.com/tendermint/spec/tree/master/spec/consensus/proposer-based-timestamp
+func (b *Block) IsTimely(clock tmtime.Source, p TimestampParams) bool {
```

Comment:
Don't we measure timeliness when we receive the proposal or when we receive the block. I thought it was based on the proposal

Reply:
In terms of the implementation, this also seems easier. If a Proposal is not timely, when it is received and processed at first by the consensus implementation, there is no point in "asking for" and receiving the corresponding BlockParts. In addition, if the timely call is made for the block, when in the implementation will it be called? This is relevant, as different calls to timely may return true and the false, depending on when they are invoked.